File: api/model/agent.go
package model

import "errors"

type Agent struct {
	AgentID           int64   `json:"agent_id" gorm:"primaryKey;autoIncrement"`
	Eid               int64   `json:"eid" gorm:"not null;index"`
	Name              string  `json:"name" gorm:"not null"`
	Logo              string  `json:"logo" gorm:"not null"`
	Sort              int     `json:"sort" gorm:"default:0"`
	Description       string  `json:"description" gorm:"not null"`
	ChannelType       int     `json:"channel_type" gorm:"default:0"`
	Model             string  `json:"model" gorm:"not null"`
	Prompt            string  `json:"prompt" gorm:"not null"`
	Configs           string  `json:"configs" gorm:"not null;type:text"`
	Tools             string  `json:"tools" gorm:"not null;type:text"`
	GroupID           int64   `json:"group_id" gorm:"type:int;default:0;not null"`
	UseCases          string  `json:"use_cases" gorm:"not null;type:text"`
	CreatedBy         int64   `json:"created_by" gorm:"not null"`
	CustomConfig      string  `json:"custom_config" gorm:"not null;type:text"`
	Settings          string  `json:"settings" gorm:"not null;type:text"`
	UserGroupIds      []int64 `json:"user_group_ids" gorm:"-"`
	Enable            bool    `json:"enable" gorm:"default:false;comment:enable status"`
	ConversationCount int64   `json:"conversation_count" gorm:"-"`
	AgentType         int     `json:"agent_type" gorm:"default:0"`
	BaseModel
}

const (
	AgentTypeApp      = 0
	AgentTypeWorkflow = 1
)

func (agent *Agent) Create() error {
	if agent.Eid == 0 {
		return errors.New("eid is empty")
	}
	// check if name exists
	var count int64
	DB.Model(&Agent{}).Where("eid = ? AND name = ?", agent.Eid, agent.Name).Count(&count)
	if count > 0 {
		return errors.New("name already exists")
	}

	result := DB.Create(agent)
	if result.Error != nil {
		return result.Error
	}

	return nil
}

func (agent *Agent) Update() error {
	err := DB.Model(agent).Updates(agent).Error
	return err
}

func (agent *Agent) Delete() error {
	err := DB.Delete(agent).Error
	return err
}

func GetAgentByID(eid int64, agentID int64) (*Agent, error) {
	var agent Agent
	err := DB.Where("eid = ? AND agent_id = ?", eid, agentID).First(&agent).Error
	if err != nil {
		return nil, err
	}
	return &agent, nil
}

func GetAgentListWithIDs(eid int64, keyword string, group_id int64, permittedAgentIDs []int64, channel_types []int, agent_types []int, offset int, limit int) (count int64, agents []*Agent, err error) {
	db := DB.Model(&Agent{}).Where("eid = ?", eid)
	if keyword != "" {
		db = db.Where("name LIKE ?", "%"+keyword+"%")
	}

	if group_id != 0 {
		db = db.Where("group_id =?", group_id)
	}

	if len(channel_types) > 0 {
		db = db.Where("channel_type IN?", channel_types)
	}

	if len(agent_types) > 0 {
		db = db.Where("agent_type IN?", agent_types)
	}

	if permittedAgentIDs != nil {
		if len(permittedAgentIDs) == 0 {
			return 0, []*Agent{}, nil
		}
		db = db.Where("agent_id IN ?", permittedAgentIDs)
	}

	db.Count(&count)

	err = db.Offset(offset).Limit(limit).Order("sort DESC").Order("agent_id DESC").Find(&agents).Error

	return count, agents, err
}

func GetAvailableAgentList(eid int64, agent_types []int, offset int, limit int) (count int64, agents []*Agent, err error) {
	db := DB.Model(&Agent{}).Where("eid = ?", eid).
		Where("Enable = ?", true)

	if len(agent_types) > 0 {
		db = db.Where("agent_type IN?", agent_types)
	}

	db.Count(&count)

	err = db.Offset(offset).Limit(limit).Order("sort DESC").Order("agent_id DESC").Find(&agents).Error

	return count, agents, err
}

func (a *Agent) GetUserGroupIds() ([]int64, error) {
	var permissions []ResourcePermission
	groupIds := make([]int64, 0)
	seen := make(map[int64]bool)

	err := DB.Where("resource_id = ? AND resource_type = ?", a.AgentID, ResourceTypeAgent).Find(&permissions).Error
	if err != nil {
		return nil, err
	}

	for _, p := range permissions {
		if !seen[p.GroupID] {
			seen[p.GroupID] = true
			groupIds = append(groupIds, p.GroupID)
		}
	}

	return groupIds, nil
}

// LoadUserGroupIds
func (a *Agent) LoadUserGroupIds() error {
	ids, err := a.GetUserGroupIds()
	if err != nil {
		return err
	}
	a.UserGroupIds = ids
	return nil
}

func (a *Agent) LoadConversationCount() error {
	var count int64
	err := DB.Model(&Conversation{}).Where("agent_id =?", a.AgentID).Count(&count).Error
	if err != nil {
		return err
	}
	a.ConversationCount = count
	return nil
}

func UpdateAgentStatus(eid, agentID int64, enable *bool) error {
	return DB.Model(&Agent{}).
		Where("eid = ? AND agent_id = ?", eid, agentID).
		Update("enable", enable).Error
}

func GetAgentCountByEID(eid int64) (int64, error) {
	var count int64
	err := DB.Model(&Agent{}).Where("eid =?", eid).Count(&count).Error
	return count, err
}

File: web/console/src/api/modules/agent.ts
import service from '../config'
import { handleError } from '../errorHandler'
import {
  AGENT_APP_OPTIONS,
  AGENT_TYPE,
  type AgentType,
} from '@/constants/platform'

export {
  AGENT_TYPE,
  AGENT_APP_OPTIONS,
  type AgentType,
}

interface AgentData {
  user_group_ids?: string | number[]
  tools?: string | any[]
  use_cases?: string | any[]
  configs?: string | Record<string, any>
  custom_config?: string | {
    agent_type?: string
    channel_config?: Record<string, any>
  }
  agent_type?: string
  agent_type_label?: string
  channel_type?: number
  enable?: boolean | number | string
  sort?: number | string
  channel_config?: Record<string, any>
  agents?: AgentData[]
  count?: number
  settings?: Record<string, any>
}

interface ListParams {
  offset?: number
  limit?: number
  keyword?: string
  group_id?: string
  channel_types?: string
}

interface SaveParams {
  agent_id?: number
  channel_type?: number
  group_id?: number
  configs?: Record<string, any> | string
  logo?: string
  name?: string
  description?: string
  model?: string
  prompt?: string
  sort?: number
  tools?: any[] | string
  use_cases?: any[] | string
  user_group_ids?: number[]
  custom_config?: Record<string, any> | string
  settings?: Record<string, any> | string
  enable?: boolean
}

interface SaveRequestData extends Omit<SaveParams, 'configs' | 'tools' | 'use_cases' | 'custom_config'> {
  configs: string
  tools: string
  use_cases: string
  custom_config: string
  settings: string
}

const parseJsonField = <T>(value: string | T, defaultValue: T): T => {
  if (typeof value !== 'string')
    return value ?? defaultValue

  try {
    return JSON.parse(value) ?? defaultValue
  }
  catch {
    return defaultValue
  }
}

export function getFormatAgentData(data: AgentData = {}): AgentData {
  data.user_group_ids = parseJsonField(data.user_group_ids, [])
  data.user_group_ids = [...new Set(data.user_group_ids)]
  data.tools = parseJsonField(data.tools, [])
  data.use_cases = parseJsonField(data.use_cases, [])
  data.configs = parseJsonField(data.configs, {})
  data.custom_config = parseJsonField(data.custom_config, {})
  data.settings = parseJsonField(data.settings, {})

  data.agent_type = data.agent_type || (data.custom_config as any)?.agent_type || AGENT_TYPE.PROMPT
  data.agent_type_label = `agent_app.${data.agent_type}`

  data.enable = !!+(data.enable ?? false)
  data.sort = +(data.sort ?? 0)
  data.channel_config = (data.custom_config as any)?.channel_config || {}
  return data
}

export const agentApi = {
  async list({ params = {} as ListParams }: { params: ListParams } = { params: {} }): Promise<AgentData> {
    params = JSON.parse(JSON.stringify(params))
    params.offset = params.offset ?? 0
    params.limit = params.limit ?? 10
    params.group_id = (!params.group_id || +params.group_id < 1) ? '0' : params.group_id

    if (!params.keyword)
      delete params.keyword

    if (!params.channel_types)
      delete params.channel_types

    const { data = {} } = await service.get('/api/agents/group', { params }).catch(handleError)
    const result = data as AgentData
    result.agents = (result.agents || []).map(item => getFormatAgentData(item))
    result.count = +(result.count ?? 0)
    return result
  },

  async save({ data = {} as SaveParams } = {}): Promise<AgentData> {
    const saveData: SaveParams = {
      agent_id: 0,
      channel_type: 0,
      group_id: 0,
      configs: {},
      logo: '',
      name: '',
      description: '',
      model: '',
      prompt: '',
      sort: 0,
      tools: [],
      use_cases: [],
      user_group_ids: [],
      custom_config: {},
      settings: {},
      enable: true,
      ...data,
    }

    const agent_id = +(saveData.agent_id ?? 0)
    delete saveData.agent_id

    const requestData: SaveRequestData = {
      ...saveData,
      configs: typeof saveData.configs === 'object' ? JSON.stringify(saveData.configs) : saveData.configs as string,
      tools: Array.isArray(saveData.tools) ? JSON.stringify(saveData.tools) : saveData.tools as string,
      use_cases: Array.isArray(saveData.use_cases) ? JSON.stringify(saveData.use_cases) : saveData.use_cases as string,
      custom_config: typeof saveData.custom_config === 'object' ? JSON.stringify(saveData.custom_config) : saveData.custom_config as string,
      settings: typeof saveData.settings === 'object' ? JSON.stringify(saveData.settings) : saveData.settings as string,
    }

    if (typeof requestData.enable === 'undefined')
      requestData.enable = true

    const { data: result = {} } = await service[agent_id ? 'put' : 'post'](`/api/agents${agent_id ? `/${agent_id}` : ''}`, requestData).catch(handleError)
    return getFormatAgentData(result)
  },

  async delete({ data: { agent_id } }: { data: { agent_id: number } }) {
    return service.delete(`/api/agents/${agent_id}`).catch(handleError)
  },

  async detail({ data: { agent_id } }: { data: { agent_id: number } }) {
    const { data = {} } = await service.get(`/api/agents/${agent_id}`).catch(handleError)
    return getFormatAgentData(data)
  },

  async updateStatus({ data: { agent_id, enable } }: { data: { agent_id: number; enable: boolean } }) {
    return service.patch(`/api/agents/${agent_id}/status`, { enable }).catch(handleError)
  },
}

export default agentApi
